contracts/integrations/utils/AdapterBase.sol

```diff
@@ -0,0 +1,178 @@
+        for (uint256 i = 0; i < incomingAssets.length; i++) {
+            require(
+                incomingAssets[i] != address(0),
+                "fundAssetsTransferHandler: incoming asset cannot be empty"
+            );
+            incomingAssetPreCallAmounts[i] = IERC20(incomingAssets[i]).balanceOf(address(this));
+        }
+
+        // Execute call
+        _;
+
+        // Transfer incoming assets back to fund
+        for (uint256 i = 0; i < incomingAssets.length; i++) {
+            uint256 postCallAmount = IERC20(incomingAssets[i]).balanceOf(address(this));
+            // Allow incoming asset balance diff to be 0 or negative in case adapters
+            // can send assets directly to the Vault
+            if (postCallAmount <= incomingAssetPreCallAmounts[i]) continue;
+
+            __safeTransfer(
+                incomingAssets[i],
+                msg.sender,
+                sub(postCallAmount, incomingAssetPreCallAmounts[i])
+            );
+        }
+
+        // If excess spend assets, send back to fund
+        for (uint256 i = 0; i < spendAssets.length; i++) {
+            uint256 postCallAmount = IERC20(spendAssets[i]).balanceOf(address(this));
+            if (postCallAmount > spentAssetPreCallAmounts[i]) {
+                __safeTransfer(
+                    spendAssets[i],
+                    msg.sender,
+                    sub(postCallAmount, spentAssetPreCallAmounts[i])
+                );
+            }
+        }
+    }
+
+    modifier onlyVault {
+        require(__isVault(msg.sender), "Only Vault can make this call");
+        _;
+    }
+
+    constructor(address _registry) public {
+        REGISTRY = _registry;
+    }
+
+    // INTERNAL FUNCTIONS
+
+    /// @dev Aggregates a list of _assets and _amounts, removing duplicate assets and summing their amounts.
+    /// At present, this is only used to aggregate fees in integration adapters.
+    /// e.g., in 0x v3, if the takerFee asset is WETH, then takerFee and protocolFee are aggregated.
+    /// It remove assets with an empty address or a balance of 0, so adapters can be dumb.
+    function __aggregateAssets(address[] memory _assets, uint256[] memory _amounts)
+        internal
+        pure
+        returns (
+            address[] memory aggregatedAssets_,
+            uint256[] memory aggregatedAssetAmounts_
+        )
+    {
+        // Get count of unique assets with non-zero values
+        uint256 aggregatedAssetsCount;
+        for (uint256 i = 0; i < _assets.length; i++) {
+            // Ignore assets with a 0 address, 0 amount
+            if (_assets[i] == address(0) || _amounts[i] == 0) continue;
+
+            // Ignore assets that have already been added
+            bool assetAdded;
+            for (uint256 j = 0; j < i; j++) {
+                // Prev asset in array must have non-zero amount
+                if (_assets[i] == _assets[j] && _amounts[j] > 0) {
+                    assetAdded = true;
```

Comment:
Is there no `continue(2)` or so in solidity?

Reply:
I don't believe so (never seen it), but the next lines act on `assetAdded` being false (i.e., having gone through the entire loop and not triggering the `true` condition), so I don't think jumping to the higher for loop is relevant here? What am I missing?